File: Leetcode/1011 Capacity To Ship Packages Within D Days.cpp
class Solution {
  public:
    int shipWithinDays(const vector<int> &weights, int days) {
        int lo = 0, hi = 0;
        // O(n) time
        for (int weight : weights) {
            lo = max(lo, weight);
            hi += weight;
        }
        int capacity = 0;
        // loop runs at most O(log s) time
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            // O(n) time
            if (canShip(mid, weights, days)) {
                capacity = mid;
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }
        // so the loop runs in O(n log s) time
        return capacity;
    }

  private:
    bool canShip(int capacity, const vector<int> &weights, int maxDays) {
        int days = 1, weightSum = 0;
        for (int weight : weights) {
            if (weightSum + weight > capacity) {
                days++;
                weightSum = 0;
            }
            weightSum += weight;
        }
        return days <= maxDays;
    }
};

File: Leetcode/1011 Capacity To Ship Packages Within D Days.py
class Solution:
    # n = length of weights, s = sum of weights
    def shipWithinDays(self, weights, days):
        lo, hi = 0, 0
        # O(n) time
        for weight in weights:
            lo = max(lo, weight)
            hi += weight
        capacity = 0
        # loop runs at most O(log s) time
        while lo <= hi:
            mid = lo + (hi - lo) // 2
            # O(n) time
            if can_ship(mid, weights, days):
                capacity = mid
                hi = mid - 1
            else:
                lo = mid + 1
        # so the loop runs in O(n log s) time
        return capacity


def can_ship(capacity, weights, max_days):
    days, weight_sum = 1, 0
    for weight in weights:
        if weight_sum + weight > capacity:
            days += 1
            weight_sum = 0
        weight_sum += weight
    return days <= max_days
